module/zfs/vdev_draid.c

```diff
@@ -0,0 +1,2962 @@
+	 */
+	vdev_open_children_subset(vd, vdev_draid_open_children);
+	vdev_open_children_subset(vd, vdev_draid_open_spares);
+
+	/* Verify enough of the children are available to continue. */
+	for (int c = 0; c < vd->vdev_children; c++) {
+		if (vd->vdev_child[c]->vdev_open_error != 0) {
+			if ((++open_errors) > nparity) {
+				vd->vdev_stat.vs_aux = VDEV_AUX_NO_REPLICAS;
+				return (SET_ERROR(ENXIO));
+			}
+		}
+	}
+
+	/*
+	 * Allocatable capacity is the sum of the space on all children less
+	 * the number of distributed spares rounded down to last full row
+	 * and then to the last full group. An additional 32MB of scratch
+	 * space is reserved at the end of each child for use by the dRAID
+	 * expansion feature.
+	 */
+	uint64_t child_asize, child_max_asize;
+	vdev_draid_calculate_asize(vd, &child_asize, &child_max_asize,
+	    logical_ashift, physical_ashift);
+
+	/*
+	 * Should be unreachable since the minimum child size is 64MB, but
+	 * we want to make sure an underflow absolutly cannot occur here.
+	 */
+	if (child_asize < VDEV_DRAID_REFLOW_RESERVE ||
+	    child_max_asize < VDEV_DRAID_REFLOW_RESERVE) {
+		return (SET_ERROR(ENXIO));
+	}
+
+	child_asize = ((child_asize - VDEV_DRAID_REFLOW_RESERVE) /
+	    VDEV_DRAID_ROWSIZE) * VDEV_DRAID_ROWSIZE;
+	child_max_asize = ((child_max_asize - VDEV_DRAID_REFLOW_RESERVE) /
+	    VDEV_DRAID_ROWSIZE) * VDEV_DRAID_ROWSIZE;
+
+	*asize = (((child_asize * vdc->vdc_ndisks) / vdc->vdc_groupsz) *
+	    vdc->vdc_groupsz);
+	*max_asize = (((child_max_asize * vdc->vdc_ndisks) / vdc->vdc_groupsz) *
+	    vdc->vdc_groupsz);
+
+	return (0);
+}
+
+/*
+ * Close a top-level dRAID vdev.
+ */
+static void
+vdev_draid_close(vdev_t *vd)
+{
+	for (int c = 0; c < vd->vdev_children; c++) {
+		if (vd->vdev_child[c] != NULL)
+			vdev_close(vd->vdev_child[c]);
+	}
+}
+
+/*
+ * Return the asize of the largest block which can be reconstructed.
+ */
+uint64_t
+vdev_draid_max_rebuildable_asize(vdev_t *vd, uint64_t max_segment)
+{
+	vdev_draid_config_t *vdc = vd->vdev_tsd;
+
+	uint64_t psize = MIN(P2ROUNDUP(max_segment * vdc->vdc_ndata,
+	    1 << vd->vdev_ashift), SPA_MAXBLOCKSIZE);
+
+	/*
+	 * When the maxpsize >> ashift does not divide evenly by the number
+	 * of data drives, the remainder must be discarded. Otherwise the skip
+	 * sectors will cause vdev_draid_asize_to_psize() to get a psize larger
+	 * than the maximum allowed block size.
+	 */
+	psize >>= vd->vdev_ashift;
+	psize /= vdc->vdc_ndata;
+	psize *= vdc->vdc_ndata;
+	psize <<= vd->vdev_ashift;
+
+	return (vdev_draid_asize(vd, psize));
+}
+
+/*
+ * Align the start of the metaslab to the group width and slightly reduce
+ * its size to a multiple of the group width.  Since full stripe write are
+ * required by dRAID this space is unallocatable.  Furthermore, aligning the
+ * metaslab start is important for vdev initialize and TRIM which both operate
+ * on metaslab boundaries which vdev_xlate() expects to be aligned.
```

Comment:
raidz always allocates in chunks of (P+1), but doesn't have corresponding metaslab-size-adjusting logic.  How does that work and why is draid handled differently?  Maybe raidz doesn't require (P+1) alignment of the absolute offset, only relative to metaslab start? and it just leaves the last bit of each metaslab unallocated?

Reply:
Since we always write full width groups with dRAID it's convenient to align the start of the metaslab with the first column of a group.  Then we always allocate chunks in multiples of the minimum allocation size (D+P).  This isn't an issue for raidz since it doesn't require anything more than sector alignment.  This does mean there's a little bit of unaligned space which spans metaslabs unused with dRAID.